packages/shared/src/lib/persisted.ts

```diff
@@ -52,3 +52,95 @@ export function persisted<T>(initial: T, key: string): Persisted<T> {
 		update
 	};
 }
+
+interface EphemeralStorageItem {
+	value: unknown;
+	expiration: number;
+}
+
+function isEphemeralStorageItem(item: unknown): item is EphemeralStorageItem {
+	const cast = item as EphemeralStorageItem;
+	return (
+		cast.value !== undefined && cast.expiration !== undefined && typeof cast.expiration === 'number'
+	);
+}
+
+function setEphemeralStorageItem(key: string, value: unknown, expiration: number): void {
+	const item: EphemeralStorageItem = {
+		value,
+		expiration
+	};
+	window.sessionStorage.setItem(key, JSON.stringify(item));
+}
+
+function getEphemeralStorageItem(key: string): unknown {
+	const item = window.sessionStorage.getItem(key);
+	try {
+		if (!item) {
+			return undefined;
+		}
+
+		const parsed = JSON.parse(item);
+		if (!isEphemeralStorageItem(parsed)) {
+			return undefined;
+		}
+
+		const now = Date.now();
+		if (parsed.expiration < now) {
+			window.sessionStorage.removeItem(key);
+			return undefined;
+		}
+
+		console.log('Item not expired:', key, parsed.value);
+
+		return parsed.value;
+	} catch {
+		return undefined;
+	}
+}
+
+/**
+ * Create a persisted store that expires after a certain amount of time.
+ */
+// eslint-disable-next-line @typescript-eslint/no-empty-object-type
+export function persistWithExpiration<T extends {}>(
```

Comment:
Allow only non-nullish values: https://www.totaltypescript.com/the-empty-object-type-in-typescript#constraining-functions